### PR TITLE
add TIME type conversion to string converter

### PR DESCRIPTION
**Pull request by p-eye**
The converter for converting STRING to DATE type is working well, 
but while converting a STRING to TIME type, an error occurred because a converter for this type does not exist. The error message is ```(NotSupportedType) cannot take column type TIME for string column.```

To fix this, I have added a converter to transform STRING type into BigQuery TIME type. I only updated the string_converter since the format "%H:%M:%S.%6N" is not compatible with the TIMESTAMP type, so I did not modify the timestamp_converter.

--- my bigquery table ---
image

--- works well with DATE column ---
```
in:
  type: inline
  schema:
    - {name: date_column, type: string}
  data:
    - {date_column: "2024-01-01"}
out:
  type: bigquery
  mode: append
  auth_method: service_account
  json_keyfile: my_keyfile
  location: asia-northeast3
  project: my_project
  destination_project: my-destination-project
  compression: GZIP
  source_format: NEWLINE_DELIMITED_JSON
  default_timezone: Asia/Seoul
  dataset: my_dataset
  table: my_table
  column_options:
  - name: date_column
    type: DATE
```

--- not works with TIME column ---
```
in:
  type: inline
  schema:
    - {name: time_column, type: string}
  data:
    - {time_column: "00:03:22"}
out:
  type: bigquery
  mode: append
  auth_method: service_account
  json_keyfile: my_keyfile
  location: asia-northeast3
  project: my-project
  destination_project: my-destination-project
  compression: GZIP
  source_format: NEWLINE_DELIMITED_JSON
  default_timezone: Asia/Seoul
  dataset: my_dataset
  table: my_table
  column_options:
  - name: time_column
    type: TIME
```

### CODE REVIEW

**Comment by p-eye** [comment]
@hiroyuki-sato
Hi, I opened a new PR. Could you please review it? 
thanks

File: lib/embulk/output/bigquery/value_converter_factory.rb
Comment (hiroyuki-sato) [inline lib/embulk/output/bigquery/value_converter_factory.rb:233]: Other converters support timestamp format and care zone offset. I think It's good to use the same way.

https://github.com/embulk/embulk-output-bigquery/blob/ee1b21d174804d8abe10d673752f07924845a209/lib/embulk/output/bigquery/value_converter_factory.rb#L192-L205

File: test/test_value_converter_factory.rb
Comment (hiroyuki-sato) [inline test/test_value_converter_factory.rb:282]: Please add some more test cases.  Like `3:15 PM`, `15:15`, `2024-07-24 10:00`.

**Review by hiroyuki-sato — CHANGES_REQUESTED** [review]
@p-eye Could you check my comments? (Thank you for proposing PR.)

File: test/test_value_converter_factory.rb
Comment (p-eye) [reply test/test_value_converter_factory.rb:282]: I added.

File: lib/embulk/output/bigquery/value_converter_factory.rb
Comment (p-eye) [reply lib/embulk/output/bigquery/value_converter_factory.rb:233]: I think the TIME type doesn't consider the offset and simply processes the time portion. So this code converts all ```2024-07-24 10:00:00 +09:00```,  ```2024-07-24 10:00:00 -09:00```, ```10:00:00``` to ```10:00:00```.

Your comment suggests that ```2024-07-24 10:00:00 +09:00``` should be converted to ```19:00:00```, and ```2024-07-24 10:00:00 -09:00``` should be converted to ```01:00:00```. And even if the timezone is ```+09:00```, the input value ```10:00:00``` should be output as ```10:00:00``` without any changes.

Is this understanding correct?

Comment (hiroyuki-sato) [reply lib/embulk/output/bigquery/value_converter_factory.rb:233]: For example, a user has a timestamp data like '2024-07-25 9:00:00 +9:00'.
For some reason, the user wants to convert the DATE and TIME fields in UTC (and also insert the same data into the DATETIME field).
In this case, data will be:
* DATE `2024-07-25`(UTC)
* DATETIME `2024-07-25 00:00:00Z`
* TIME: `00:00:00`(UTC)

If a user wants to insert in the `+9:00` timezone,
* DATE `2024-07-25`(+09:00)
* DATETIME `2024-07-25 09:00:00` (+09:00)
* TIME: `15:00:00` (+09:00)

IIUC, DATA and DATETIME fields care about that. so I think It is necessary to care about the `TIME` fields too.

Does this PR work like this? If, so I'll check like this PR.

[Attached codes](https://github.com/embulk/embulk-output-bigquery/pull/168/files#r1689748990) was just for example, It will be `TimeWithZone.set_zone_offset(Time.strptime(val, @timestamp_format), zone_offset).strftime("%H:%M:%S.%6N")`

Comment (p-eye) [reply lib/embulk/output/bigquery/value_converter_factory.rb:233]: i fixed, but test case that was previously validated successfully has now failed..
Could you check the code?

Comment (hiroyuki-sato) [reply lib/embulk/output/bigquery/value_converter_factory.rb:233]: I will check the code. the reason for the error is you set the timestamp format explicitly. This code expects the input data format `%H:%M:%S`.

```ruby
          converter = ValueConverterFactory.new(
            SCHEMA_TYPE, 'TIME',
            timestamp_format: '%H:%M:%S'
          ).create_converter
```

Comment (p-eye) [reply lib/embulk/output/bigquery/value_converter_factory.rb:233]: Thanks, I removed the type constraint from the test cases input. 
And I added support for the Time type in the timestamp converter after I got your example of converting '2024-07-25 9:00:00 +9:00'. 😃

Comment (hiroyuki-sato) [reply lib/embulk/output/bigquery/value_converter_factory.rb:233]: Thanks! Please wait a while, I need to understand the timestamp behavior of this plugin.
I don't understand why `TIMESTAMP` uses `TimeWithZone.set_zone_offset` and `DATETIME` doesn't use it yet.

https://github.com/embulk/embulk-output-bigquery/blob/8e7df4a8ea638a8cc5a77e991e967d40e5c6e2a4/lib/embulk/output/bigquery/value_converter_factory.rb#L192-L226

Comment (p-eye) [reply lib/embulk/output/bigquery/value_converter_factory.rb:233]: I agree with your suggestion that ```DATETIME``` should use ```TimeWithZone.set_zone_offset```. For the same reason, I think the ```TIME``` code might also be changed in the same way.

```
          when 'DATETIME'
            if @timestamp_format
              Proc.new {|val|
                next nil if val.nil?
                with_typecast_error(val) do |val|
                  TimeWithZone.set_zone_offset(Time.parse(val), zone_offset).strftime("%Y-%m-%d %H:%M:%S.%6N")
                end
              }
            else
              Proc.new {|val|
                next nil if val.nil?
                val # Users must care of BQ timestamp format
              }
            end
          when 'TIME'
             if @timestamp_format
              Proc.new {|val|
                next nil if val.nil?
                with_typecast_error(val) do |val|
                  TimeWithZone.set_zone_offset(Time.parse(val), zone_offset).strftime("%H:%M:%S.%6N")
                end
              }
            else
              Proc.new {|val|
                next nil if val.nil?
                val # Users must care of BQ timestamp format
              }
            end
```

**Comment by dp-daesung** [comment]
This is exactly what I need. I'm waiting for it to merge.

File: lib/embulk/output/bigquery/value_converter_factory.rb
Comment (hiroyuki-sato) [reply lib/embulk/output/bigquery/value_converter_factory.rb:233]: @p-eye Thank you for waiting. It took a long time.

I misunderstood the `TimeWithZone` feature
https://sonots.github.io/time_with_zone/TimeWithZone.html
IIUC, `TimeWithZone`.zone_offset only changes `timezone` field.

I'm not 100% sure, but I think It is not meaningful to use `TimeWithZone` in the time part.
So, We can choose (The result will be the same whether it is there or not.)

1. Do not use `TimeWithZoe.zone_offset` and add the comment `TimeWithZone doesn't affect any change`. So, it is not used in this part.
2. Add `TimeWithZoe.zone_offset` for consistency. and add  `TimeWithZone doesn't affect any change`. So, Use consistency with other parts.

My opinion is No.2.

```
[25] pry(main)> TimeWithZone.set_zone_offset(Time.strptime("15:00:01", "%H:%M:%S"), TimeWithZone.zone_offset("Asia/Tokyo")).strftime("%H:%M:%S.%6N")
=> "15:00:01.000000"
[26] pry(main)> TimeWithZone.set_zone_offset(Time.strptime("15:00:01", "%H:%M:%S"), TimeWithZone.zone_offset("UTC")).strftime("%H:%M:%S.%6N")
=> "15:00:01.000000"
```

`2024-08-10 15:00:01 +0900` is "+9:00" offset.
It is `2024-08-10 14:00:01 +0800` in `Asia/Thaipei` timezone.
But `TimeWithZone` returns `2024-08-10 15:00:01.000000 +08:00`

```
[27] pry(main)> TimeWithZone.set_zone_offset(Time.strptime("2024-08-10 15:00:01 +0000", "%Y-%m-%d %H:%M:%S %z"), TimeWithZone.zone_offset("UTC")).strftime("%Y-%m-%d %H:%M:%S.%6N %:z")
=> "2024-08-10 15:00:01.000000 +00:00"
[29] pry(main)> TimeWithZone.set_zone_offset(Time.strptime("2024-08-10 15:00:01 +0900", "%Y-%m-%d %H:%M:%S %z"), TimeWithZone.zone_offset("Asia/Taipei")).strftime("%Y-%m-%d %H:%M:%S.%6N %:z")
=> "2024-08-10 15:00:01.000000 +08:00"
```

What do you think?

Comment (p-eye) [reply lib/embulk/output/bigquery/value_converter_factory.rb:233]: yes, i agree with the No.2 idea.

Comment (hiroyuki-sato) [reply lib/embulk/output/bigquery/value_converter_factory.rb:233]: @p-eye  Thanks. Please add `TimeWithZoe.zone_offset` (and comment) in `TIME`. And, keep `DATETIME` as is. (I would better to add `TimeWithZoe.zone_offset`. But this code deploys for a long time I don't want to care about unexpected behavior.)

Comment (p-eye) [reply lib/embulk/output/bigquery/value_converter_factory.rb:233]: i fixed. please check https://github.com/embulk/embulk-output-bigquery/pull/168/commits/489b6b2a1650878983db77f3bdab46e411e0b227. thanks

File: test/test_value_converter_factory.rb
Comment (hiroyuki-sato) [inline test/test_value_converter_factory.rb:384]: Remove the trailing spaces.

**Review by hiroyuki-sato — COMMENTED** [review]
Thanks. I'll approve after removing trailing spaces.

File: test/test_value_converter_factory.rb
Comment (p-eye) [reply test/test_value_converter_factory.rb:384]: Sorry, did you mean the blank lines (lines 376 and 384)? I can't find any trailing spaces on each line.

Comment (hiroyuki-sato) [reply test/test_value_converter_factory.rb:384]: Line 278

```diff
diff --git a/test/test_value_converter_factory.rb b/test/test_value_converter_factory.rb
index c49ef07..89c31c1 100644
--- a/test/test_value_converter_factory.rb
+++ b/test/test_value_converter_factory.rb
@@ -275,7 +275,7 @@ module Embulk
             SCHEMA_TYPE, 'TIME', timezone: 'Asia/Tokyo'
           ).create_converter
           assert_equal "15:00:01.000000", converter.call("15:00:01")
-          
+
           assert_raise { converter.call('foo') }
         end
 
```

Comment (p-eye) [reply test/test_value_converter_factory.rb:384]: thanks, i removed it

**Review by hiroyuki-sato — APPROVED** [review]
@p-eye  LGTM 👍 Thanks! @joker1007 Could you take a look when you get a chance?

**Comment by p-eye** [comment]
hi, any updates? 😃

**Comment by hiroyuki-sato** [comment]
@p-eye I pinged reviewer. The maintainer will review it soon. Please wait a bit

**Review by joker1007 — APPROVED** [review]
I understand. LGTM.

**Comment by hiroyuki-sato** [comment]
Thanks!

**Comment by hiroyuki-sato** [comment]
@p-eye , @dp-daesung `embulk-output-bigquery` 0.7.3 has been released. Enjoy.
@p-eye Thank you for your contribution.

**Pull request merged**
